mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/transformation/cache/MpsToModelixMap.kt

```diff
@@ -227,6 +231,225 @@ object MpsToModelixMap {
     fun isMappedToModelix(module: SModule) = this[module] != null
 
     fun isMappedToModelix(node: SNode) = this[node] != null
+
+    fun isEmpty() = objectsRelatedToAModel.isEmpty() && objectsRelatedToAModule.isEmpty()
+
+    fun clear() {
+        nodeToModelixId.clear()
+        modelixIdToNode.clear()
+        modelToModelixId.clear()
+        modelixIdToModel.clear()
+        moduleToModelixId.clear()
+        modelixIdToModule.clear()
+        moduleWithOutgoingModuleReferenceToModelixId.clear()
+        modelixIdToModuleWithOutgoingModuleReference.clear()
+        modelWithOutgoingModuleReferenceToModelixId.clear()
+        modelixIdToModelWithOutgoingModuleReference.clear()
+        modelWithOutgoingModelReferenceToModelixId.clear()
+        modelixIdToModelWithOutgoingModelReference.clear()
+        objectsRelatedToAModel.clear()
+        objectsRelatedToAModule.clear()
+    }
+
+    class Serializer {
```

Comment:
fixed in 8aecd55, daa5556 and defbe0b

Reply:
After removing the JSON serialization completely, the traversal-based solution is implemented in [12daf9a](https://github.com/modelix/modelix.mps-plugins/pull/94/commits/12daf9a63a14cc5d67227504f87ce36fd81b2caf) and refactored in [2157383](https://github.com/modelix/modelix.mps-plugins/pull/94/commits/21573835d06914c3fc9efd57f9b5d89781dac41b). It is still in a WIP phase, however it seems to work fine for the courses example. (Working fine = the join points between the coroutines are kept, so parallelism / concurrency happens only where we expect that.) I'll test some error-handling scenarios before being finished with the task. Plus document some new public APIs (e.g. the Visitor). 

The solution is similar to how the `ITreeToSTreeTransformer` and the `ModuleTransformer`,  `ModelTransformer`, `NodeTransformer` works. The biggest difference is, instead of `ContinuableSyncTasks`, `SyncQueue` and `CompletableFutures`, I used only coroutines with children coroutine scopes. So it might be more efficient and less error-prone than running Runnables in ThreadPools (`Executors`).

The reason for implementing it like this was, the model traversal can be parallelized where it makes sense so we do not have to wait for a single thread to complete it. Since `ITreeToSTreeTransformer` goes in the same direction, I thought it makes sense to do the same, but now with coroutines so that we can how a coroutine-based solution would look like (as it was already suggested in [another code review earlier](https://github.com/modelix/modelix.mps-plugins/pull/85#discussion_r1570689570)).

Not all corner cases are implemented in the `ITreeTraversal` and the `MpsToModelixMapInitializerVisitor`. For this use case we just need read actions in MPS and read transactions in modelix. However, in the *Transformer use cases we have use cases where a write action has to wait for the result of another write action. We shall see how we can implement it in the coroutine-based solution.

If the coroutine-based solution seems to work fine, then we can add it as a technical debt to [MODELIX-637](https://issues.modelix.org/issue/MODELIX-637). It had been there previously but was erased, because we did not have a reference (or starting point?) implementation. Now, we have it.